Clarify text around preferred address (#3589)
quicwg/base-drafts: draft-ietf-quic-transport.md

be used again and requests that the peer replace it with a new connection ID
using a NEW_CONNECTION_ID frame.

As discussed in {{migration-linkability}}, endpoints MUST limit the use of a
connection ID to packets sent from a single local address.  Endpoints SHOULD
retire connection IDs when they are no longer actively using the network path on

> which the connection ID was used.

Reviewer: This text is now contradictory, because I can send from A -> B and A->C but the subsequent text now refers to "the network path"

Reviewer: Perhaps the problem is the use of "network path". As I understand it, the requirement here is that from your perspective the 5-tuple be the same, right? The issue is that with NAT your perspective and others may different. So  perhaps "MUST limit the use of a connection ID to a single pair of remote/local address"

Author: > Perhaps the problem is the use of "network path".

I would agree, I think we've been trying to steer away from "network path" in favor of things from the perspective of each endpoint.

> The point about NAT rebinding/inadvertent migration here is not relevant and somewhat distracting.

The NAT rebinding is something that's been covered elsewhere already, I believe, but every review that we've had of this text has resulted in someone raising their hand to say "don't forget about NAT rebinding here", so have kept text in to settle that.

> The issue is that with NAT your perspective and others may different.

When we distilled this down, we went from the perspective of the requirements that are placed on each endpoint to help cope with this.

> This text is now contradictory, because I can send from A -> B and A->C

That shouldn't be the case: 
```
An endpoint MUST NOT reuse a connection ID when sending from more than one local
address, ...
```
```
Similarly, an endpoint MUST NOT reuse a connection ID when sending to more than
one destination address.
```

I think we can change the text above to not mention a "network path" and instead talk about the destination address, since that's what we really mean in this case.

Author: Updated the text in f7a9c3b to refer to being done with either your local address (migrated to a new address) or your destination address (only relevant for preferred_address and someday maybe server migration). Whenever you're done with an address on either end, you should probably retire the CID that you used with that address.

be used again and requests that the peer replace it with a new connection ID
using a NEW_CONNECTION_ID frame.

> As discussed in {{migration-linkability}}, endpoints MUST limit the use of a

Reviewer 2: ```suggestion
As discussed in {{migration-linkability}}, endpoints limit the use of a
```
There is an exception in the referenced text that isn't repeated here.  We don't need to repeat that, so it is best to keep this as advisory text.

At any time, endpoints MAY change the Destination Connection ID they send to a
value that has not been used on another path.

An endpoint MUST NOT reuse a connection ID when sending from more than one local
address, for example when initiating connection migration as described in
{{initiating-migration}} or when probing a new network path as described in
{{probing}}.

> Similarly, an endpoint MUST NOT reuse a connection ID when sending to more than
> one destination address, for example when responding to a change in the address
> of a peer if the packet with the new peer address uses an active connection ID
> that has not been previously used by the peer.

Reviewer 2: This is a little long and the 'for example' is a little buried.

```suggestion
Similarly, an endpoint MUST NOT reuse a connection ID when sending to a remote
address from which the peer has chosen to use a connection ID that has not
previously been used.  This occurs when an endpoint receives packets that
include a change to both the source address and destination connection ID at the
same time.
```

Reviewer 2: I'm happy with Mike's suggestion.

Reviewer 3: Is rationale useful here?   ie "as this indicates the peer did not intentionally change its address."?

Author: Thanks! This is nice wording, added in 5509f1d.
 
Re: rationale, I think that's mostly covered by two notes indicating that this can happen.
This one, the immediate next paragraph after the sentence in question:
```
These requirements regarding connection ID reuse apply only to the sending of
packets, as unintentional changes in path without a change in connection ID are
possible.  For example, after a period of network inactivity, NAT rebinding
might cause packets to be sent on a new path when the client resumes sending.
An endpoint responds to such an event as described in {{migration-response}}.
```
and, immediately after the other place we talk about this kind of interaction: 
```
Not all changes of peer address are intentional, or active, migrations. The peer
could experience NAT rebinding: a change of address due to a middlebox, usually
a NAT, allocating a new outgoing port or even a new outgoing IP address for a
flow.  An endpoint MUST perform path validation ({{migrate-validate}}) if it
detects any change to a peer's address, unless it has previously validated that
address.
```

> Note that these requirements apply to the endpoints sending packets, as
> unintentional changes in path without a change in connection ID are possible.
> For example, after a period of network inactivity, NAT rebinding might cause
> packets to be sent on a new path when the client resumes sending.  An endpoint
> responds to such an event as described in {{migration-response}}.

Reviewer 2: ```suggestion
These requirements regarding connection ID reuse apply only to the sending of
packets, as unintentional changes in path without a change in connection ID are
possible.  For example, after a period of network inactivity, NAT rebinding
might cause packets to be sent on a new path when the client resumes sending.
An endpoint responds to such an event as described in {{migration-response}}.
```

At any time, endpoints MAY change the Destination Connection ID they send to a
value that has not been used on another path.

An endpoint MUST NOT reuse a connection ID when sending from more than one local
address, for example when initiating connection migration as described in
{{initiating-migration}} or when probing a new network path as described in
{{probing}}.

Similarly, an endpoint MUST NOT reuse a connection ID when sending to more than
one destination address.  However, if an endpoint receives packets from a new
source address with the same destination connection ID, it MAY continue to use

> the current connection ID with the new address.

Reviewer: This text is really confusing. We have a direct prohibition followed by an unexplained exception. Can we provide some reasoning here?

Reviewer 2: How about:

> However, if a peer experiences NAT rebinding, an endpoint could receive packets from its peer using a new source address with the same destination connection ID.  If a peer is first to use a connection ID on a new network path and that connection ID was originally used on another network path, an endpoint MAY also use a connection ID that it used on the original network path.

That is much more wordy, but it might be clearer.  The point is to say that you don't reuse a connection ID unless your peer reuses first.

There is a wrinkle here in the case that an endpoint has multiple valid addresses (think preferred address).  In that case, we don't want to allow reuse as that implies a bunch of other things.  The fix there might be to add:

> This exception that allows the use of a connection ID on multiple network paths only applies when the remote address changes and the local address remains constant.

Reviewer: Do we actually have to make this exception? I haven't been tracking all of the conversations, but why not just have endpoints use a new CID in this case.

In general an endpoint MUST limit the use of a connection ID to a single pair of remote and local addresses. This prevents linkage between network paths. However, if an endpoint may experiences unexpected NAT rebinding, that endpoint's peer may receive a packet from a new remote address but reusing an existing connection ID, thus apparently violating this rule linking those paths. [TODO: Something about how you have to accept this?]
As an exception to the rule above, if an endpoint sees a given connection ID reused from a new remote address, it MAY also reuse its corresponding connection ID when sending to that address, as long as it sends from the same local address.

Author: I like the framing of the last sentence you have there, although it's interesting to think through the question about if we really need an exception here. 

In theory, the problem with changing for this is that a NAT rebinding can cause us to "burn" a CID, making it so that instead of having 2 linked CIDs, we now have 3 linked CIDs. As long as we retire the old one (which is already in the text) once we change to the new one, we should be able to keep a steady supply of those around, although it would mean that someone in the middle of the network could fiddle with things to keep us generating and exchanging more and more CIDs. 

So, the idea is that if someone on the network changes without your knowledge, we've already linked those two 5-tuples by the fact that you've got the same CID in one direction, there's no harm in continuing to keep using that CID in the other direction. 

It does seem better to keep doing that, rather than opening ourselves up to letting someone on the network arbitrarily churn through CIDs.

This would mean the text becomes much like what you have in the TODO above, essentially we're saying that the _sender_ of the packet MUST follow these restrictions, but if you see that your peer did not, then \<the text you have above\>.

Reviewer 2: I am comfortable with either outcome here.  @erickinnear, do you want to try to integrate these comments in something approximating the shape you describe?

Author: Yup! Working on an update for these this afternoon.

Author: Updated to fold this in, also note that much of this explanation (and further requirements on behavior in the event of a NAT rebinding, such as potentially needing to validate the new path) is present in the immediately following paragraph: 
```
These requirements regarding connection ID reuse apply only to the sending of
packets, as unintentional changes in path without a change in connection ID are
possible.  For example, after a period of network inactivity, NAT rebinding
might cause packets to be sent on a new path when the client resumes sending.
An endpoint responds to such an event as described in {{migration-response}}.
```
Updated in f7a9c3b

Reviewer 2: Thanks for doing that Eric.  It looks good to me.

At any time, endpoints MAY change the Destination Connection ID they send to a
value that has not been used on another path.

An endpoint MUST NOT reuse a connection ID when sending from more than one local
address, for example when initiating connection migration as described in
{{initiating-migration}} or when probing a new network path as described in
{{probing}}.

> Similarly, an endpoint MUST NOT reuse a connection ID when sending to more than
> one destination address.  Due to network changes outside the control of its
> peer, an endpoint might receive packets from a new source address with the same
> destination connection ID, in which case it MAY continue to use the current
> connection ID with the new remote address while still sending from the same
> local address.

Reviewer: ```suggestion
Due to network changes outside the control of its
peer, an endpoint might receive packets from a new source address with the same
destination connection ID, in which case it MAY continue to use the current
connection ID with the new remote address while still sending from the same
local address. In all other cases, an endpoint MUST NOT reuse a connection ID when sending to more than
one destination address. 
```

Author: I don't dislike the flow once flipped, but the first sentences of this and the preceding paragraph are trying to be very clearly the main requirement, with the subsequent items explaining motivation and any exceptions.

This is in response to the original issue we're trying to address, which is that all of these requirements were in place but got lost in the rest of the text, so the idea was to highlight the main requirements in a single place.

Thoughts?

Reviewer 2: I understand the spirit of the change, which is to make the primary requirement properly conditional, but it does so by burying that requirement.  I'm comfortable with the original.  Happy to iterate again later.

Servers MAY communicate a preferred address of each address family (IPv4 and
IPv6) to allow clients to pick the one most suited to their network attachment.

Once the handshake is confirmed, the client SHOULD select one of the two
server's preferred addresses and initiate path validation (see
{{migrate-validate}}) of that address using the connection ID provided in the
preferred_address transport parameter or, if that connection ID has been
retired, any active connection ID that has not already been used on a different

> path.

Reviewer 2: I can't suggest here, but maybe "using any previously unused connection ID, taken from either the preferred_address transport parameter or the NEW_CONNECTION_ID frame."

Reviewer 4: I prefer @martinthomson's suggestion. For CIDs provided by NCID frames, we do not require them to be used in particular order, even when the client has received CIDs in order. I would be hesitant to to require in-order use of CID when SPA is involved.

Author: Thanks! I was struggling to find a way to word that nicely, this is much better.

Author: Fixed in 5509f1d

Retire Prior To:

: A variable-length integer indicating which connection IDs should be retired.

>   See {{retire-cid}}.

Reviewer 2: semi-colon here too.

Author: Done in 3cf8eaa

PROTOCOL_VIOLATION.

The Retire Prior To field counts connection IDs established during connection

> setup and the preferred_address transport parameter (see {{retire-cid}}). The

Reviewer 2: ```suggestion
setup and the preferred_address transport parameter; see {{retire-cid}}. The
```

Reviewer 3: I thought parens were fine?  Personally, I prefer parens, so I'm curious why prefer semicolons?

I think they're used quite widely, so if you want to consistently move to semicolons, I'd suggest a single cleanup PR to switch everything over.

Sequence Number:

> : The sequence number of the connection ID being retired.  See

Reviewer 5: ```suggestion
: The sequence number of the connection ID being retired; see
```

>   {{retire-cid}}.

Reviewer 2: I can't suggest here (damn GitHub).  But what you want is a semi-colon: '; see {{retire-cid}}.'

Author: Done in 3cf8eaa
